File: pkg/sign/sign.go
package sign

import (
	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
	hash "github.com/bsv-blockchain/go-sdk/primitives/hash"
	v1 "github.com/spycat55/KeymasterProto/gen/go"
	encodepkg "github.com/spycat55/KeymasterProto/pkg/encode"
	"google.golang.org/protobuf/proto"
)

// SignEnvelope calculates signature over deterministic marshal of the envelope
// with Signature fields cleared and fills Signature & SignatureAlgo.
func SignEnvelope(env *v1.Envelope, priv *ec.PrivateKey) error {
	// clone and clear sig
	clone := proto.Clone(env).(*v1.Envelope)
	clone.Signature = nil
	clone.SignatureAlgo = ""

	b, err := encodepkg.DeterministicMarshal(clone)
	if err != nil {
		return err
	}
	digest := hash.Sha256(b)

	sig, err := priv.Sign(digest)
	if err != nil {
		return err
	}
	der, err := sig.ToDER()
	if err != nil {
		return err
	}
	env.Signature = der
	env.SignatureAlgo = "ECDSA_SECP256K1_SHA256"
	return nil
}

File: packages/utils/src/sign.ts
// @ts-nocheck
/* eslint-disable */
import { deterministicMarshal } from "./encode";
import { Envelope } from "@spycat55/keymaster_proto";
import PrivateKey from "@bsv/sdk/primitives/PrivateKey";
import { sha256 } from "@bsv/sdk/primitives/Hash";
import { Buffer } from "buffer";
import { sign as ecdsaSign } from "@bsv/sdk/primitives/ECDSA";
import { toHex } from "@bsv/sdk/primitives/utils";
import BigNumber from "@bsv/sdk/primitives/BigNumber";

export function signEnvelope(env: Envelope, priv: PrivateKey): void {
  const clone: Envelope = { ...env, signature: Buffer.alloc(0), signatureAlgo: "" } as Envelope;
  const sigInput = deterministicMarshal(clone);
  const digest = sha256(sigInput);
  const digestBN = new BigNumber(toHex(digest), 16);
  const sigObj = ecdsaSign(digestBN, priv, true);
  const der = sigObj.toDER();
  env.signature = Buffer.from(der);
  env.signatureAlgo = "ECDSA_SECP256K1_SHA256";
} 